midterm-2023-11-06/code/5029-angel-nineski/script.ts

```diff
@@ -0,0 +1,184 @@
+    });
+}
+
+// Function to sort the table data
+function sortTable(bands: Band[], sortBy: string) {
+    bands.sort((a, b) => {
+        let aValue, bValue: number | string;
+
+        switch (sortBy) {
+            case "id":
+                aValue = a.id;
+                bValue = b.id;
+                break;
+            case "name":
+            case "location":
+            case "country":
+                aValue = a[sortBy];
+                bValue = b[sortBy];
+                break;
+            case "formed":
+            case "firstAlbumYear":
+            case "lastAlbumYear":
+                aValue = a[sortBy];
+                bValue = b[sortBy];
+                break;
+            case "yearsActive":
+                const aStartYear = a.albums[0].year;
+                const aEndYear = a.albums[a.albums.length - 1].year;
+                const bStartYear = b.albums[0].year;
+                const bEndYear = b.albums[b.albums.length - 1].year;
+                return aStartYear - bStartYear || aEndYear - bEndYear;
+            default:
+                aValue = 0;
+                bValue = 0;
+        }
+
+        if (typeof aValue === 'string' && typeof bValue === 'string') {
+            return (aValue as string).localeCompare(bValue as string);
+        } else if (typeof aValue === 'number' && typeof bValue === 'number') {
+            return (aValue as number) - (bValue as number);
+        } else {
+            return 0;
+        }
+    });
+}
+
+// Function to populate filter options
+function populateFilterOptions(data: any) {
+    const countryFilter = document.getElementById("country-filter");
+    const genreFilter = document.getElementById("genre-filter");
+
+    const countries = new Set<string>();
+    const genres = new Set<string>();
+
+    data.metalBands.forEach((band: Band) => {
+        countries.add(band.country);
+        genres.add(band.genre);
+    });
+
+    countries.forEach((country) => {
+        const option = document.createElement("option");
+        option.value = country;
+        option.textContent = country;
+        countryFilter?.appendChild(option);
+    });
+
+    genres.forEach((genre) => {
+        const option = document.createElement("option");
+        option.value = genre;
+        option.textContent = genre;
+        genreFilter?.appendChild(option);
+    });
+}
+
+// Function to apply filters
+function applyFilters() {
+    const nameFilter = (document.getElementById("name-filter") as HTMLInputElement).value.toLowerCase();
+    const countryFilter = (document.getElementById("country-filter") as HTMLSelectElement).value;
+    const genreFilter = (document.getElementById("genre-filter") as HTMLSelectElement).value;
+    const yearActiveFilter = parseInt((document.getElementById("year-active-filter") as HTMLInputElement).value, 10);
+
+    const filteredBands = bands.filter((band) => {
+        const nameMatches = !nameFilter || band.name.toLowerCase().includes(nameFilter);
```

Comment:
One big method for the filter again :) These are way more problematic for debugging, but it works. +15